mathesar_ui/src/stores/table-data/constraints.ts

```diff
@@ -95,6 +120,92 @@ export class ConstraintsDataStore implements Writable<ConstraintsData> {
     return null;
   }
 
+  async add(constraintDetails: Partial<Constraint>): Promise<Partial<Constraint>> {
+    const constraint = await this.api.add(constraintDetails);
+    await this.fetch();
+    return constraint;
+  }
+
+  async remove(constraintId: number): Promise<void> {
+    await this.api.remove(constraintId);
+    await this.fetch();
+  }
+
+  /**
+   * A constraint only matches if the set of its columns strictly equals the set
+   * of columns supplied here. For example, if a constraint is set on three
+   * columns, two of which are passed to this function, that constraint will
+   * _not_ be returned.
+   */
+  constraintsThatMatchSetOfColumns(columns: Column[]): Readable<Constraint[]> {
+    const columnsNames = columns.map((c) => c.name);
+    function isMatch(constraint: Constraint) {
+      if (constraint.columns.length !== columnsNames.length) {
+        return false;
+      }
+      return constraint.columns.every(
+        (constraintColumn) => columnsNames.includes(constraintColumn),
+      );
+    }
+    return derived(this.store, (s) => s.constraints.filter(isMatch));
+  }
+
+  /**
+   * Caveat: even though primary key columns must be unique, this function will
+   * give `false` for them because they don't usually have unique constraints
+   * set too.
+   */
+  columnHasUniqueConstraint(column: Column): Readable<boolean> {
+    const constraints = this.constraintsThatMatchSetOfColumns([column]);
+    return derived(constraints, (c) => c.some((constraint) => constraint.type === 'unique'));
+  }
+
+  columnAllowsDuplicates(column: Column): Readable<boolean> {
+    return derived(
+      this.columnHasUniqueConstraint(column),
+      (hasUniqueConstraint) => {
+        if (column.primary_key) {
+          return false;
+        }
+        return !hasUniqueConstraint;
+      },
+    );
+  }
+
+  async updateUniquenessOfColumn(
+    column: Column,
+    updater: (isUnique: boolean) => boolean,
```

Comment:
What is the reasoning behind this updater argument?

Since this function only requires a boolean to identify whether or not to update the uniqueness, wouldn't a boolean suffice here?

Reply:
Yeah good question. I remember having a reason for using an updater -- but I can't remember what the reason was! I'm guessing that some other changes I made along the way have now rendered my original rationale moot. In any case, I've made a change in bebb4e9372515fcb6b17a7f2ac0edb88bb9f3180 that simplified this function to use a value instead of an updater. Thanks for pointing this out.